### ISSUE TITLE
Support custom QuickPick filter logic

### ISSUE THREAD

**Issue by brunnerh** [microsoft/vscode#90521]
Currently the `QuickPick` filter logic supports wild cards and which parts of the item to search in but not things like switched up word order or regex. [An issue](https://github.com/brunnerh/insert-unicode/issues/13) was opened for one of my extensions asking for a more lenient search, which does not seem possible with the current API.

Rather than adding a lot of options specifying exactly how the filtering behaves it might be easier and more sensible to provide a custom filter callback and maybe a filter delay option. The delay would trigger the filter logic only `n` milliseconds after the user has last typed, in case the filter logic is complex or there are many items (- i have 33k -) causing long filter times.

Suggested interface for the options:

```ts
interface QuickPickOptions {
	// ...

	/**
	 * Time in milliseconds to wait after last user input
	 * before filtering the list.
	 * Default: 0
	 */
	filterDelay: number,

	/**
	 * Optional filter callback which overrides the default filter
	 * behavior. The callback is called for every item in the list
	 * with the current user input.
	 * Default: undefined
	 */
	filterCallback?: (item: QuickPickItem | string, filterText: string) => boolean,
}
```

I had a look at handling a quick pick "manually" by using `createQuickPick` but it also filters completely automatically.

---

Maybe `QuickPickOptions` could also be made generic. It already has the `onDidSelectItem` property which uses a union for the item type instead of a generic. Then functions like `showQuickPick<T extends QuickPickItem>` could propagate their generic type to the options argument.

**Comment by rmanthorpe** [comment]
@brunnerh if we could disable `matchOnLabel` (see #83425) you could just disable matching and implement the filtering yourself before you put the `QuickPickItem`s into the `QuickPick`. Coupled with custom highlighting (#83424)  you would be able to do whatever you like.

**Comment by tatosjb** [comment]
I agree with @brunnerh, about disable match **at all**, so we could handle the filter ourseves

**Comment by chrmarti** [comment]
Setting `alwaysShow: true` on the items and `sortByLabel: false` (#73904 proposed API) on the QuickPick allows you to control the list and sorting. Only the highlighting cannot be controlled yet, that is tracked with #83424.

Keeping this issue open as its own proposal until the issues are resolved.

**Comment by tatosjb** [comment]
The `alwaysShow: true` does nos seems to work this way. I have tried to use, but it still keep hiding results based on it owns filter algorithm. 

Because of this, i have to do an workaround  here:
https://github.com/tatosjb/vscode-fuzzy-search/blob/master/src/search.ts#L66

`${this.searchString} - ${path.parse(filePath).dir.replace(workspacePath || '', '')}` the `${this.searchString} - ` on the start of the expression is necessary because of that.

**Comment by eamodio** [comment]
Another option, simpler but less flexible, would be to add another prop, like `filterDetail: string` and `matchOnFilterDetail`, which would work the same as `matchOnDetail` but just not be visible, but could play along with the other options.

The rationale here, is that I often don't want to overload the quickpick item with visual noise just to match against.

Although if something more custom was implemented, I would love it if it was a post-process on top of the existing matching (if enabled -- and it would be great to be able to opt-out of label matching), so that way if something matched because of the existing behavior the item would still be there (unless removed as part of this callback) and highlighting would still be intact on those items.

**Comment by lionel-rowe** [comment]
Are there alternative, lower-level UI components that could be used instead of `QuickPick`? Input noticeably lags on my machine, which could be mitigated by running the search logic in workers, or throttling input, or virtualizing the list elements (if that isn't done already), or some combination of the above. It seems like `QuickPick` isn't really intended for searching through tens of thousands of items as it doesn't have these customization options.

As well as performance and fuzziness, matching by alternate names would also be helpful, e.g. I expected to see `0x7d MULTIPLICATION SIGN` as an option when searching for `times`.

**Comment by ArturoDent** [comment]
Second @eamodio's suggestion of the ability to match on something not shown, like `Detail` or `Description`).  My use case is an extension that uses `tsc` to get the nodes of a file for a QuickPick (including arguments, calls and parameters).  I would like the user to be able to filter for `function call` (or `constructor`, etc., etc.) and get just those without having to put `function call` into the `Description` or `Detail` - which would otherwise be very cluttered as a result.
